
#### NOTE: _All three parts of the assignment are included in this single notebook._

# Part 1:

In [1]:
import pandas as pd
import requests

In [4]:
#getting data from url.

url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
result = requests.get(url)
result      #response 200 will mean connection successfull.

<Response [200]>

In [6]:
data=pd.read_html(result.text)
data

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 ..          ...               ...   
 175         M5Z      Not assigned   
 176         M6Z      Not assigned   
 177         M7Z      Not assigned   
 178         M8Z         Etobicoke   
 179         M9Z      Not assigned   
 
                                          Neighbourhood  
 0                                         Not assigned  
 1                                         Not assigned  
 2                                            Parkwoods  
 3                                     Victoria Village  
 4                            Regent Park, Harbourfront  
 ..                                                 ...  
 175                                       Not assigned  
 176                                       Not assigned  
 177                

In [7]:
data=data[0]     #we need the first table only.
data

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

In [9]:
#dropping those boroughs which are not assigned.
df = data[data['Borough'] != "Not assigned"]
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [12]:
#grouping those areas having same postal code.
df.groupby(['Postal Code']).head()
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [13]:
#to check where neighbourhood is not assigned
df.Neighbourhood.str.count("Not assigned").sum()

0

In [15]:
df = df.reset_index()
df

level_0  index Postal Code           Borough  \
0          0      2         M3A        North York   
1          1      3         M4A        North York   
2          2      4         M5A  Downtown Toronto   
3          3      5         M6A        North York   
4          4      6         M7A  Downtown Toronto   
..       ...    ...         ...               ...   
98        98    160         M8X         Etobicoke   
99        99    165         M4Y  Downtown Toronto   
100      100    168         M7Y      East Toronto   
101      101    169         M8Y         Etobicoke   
102      102    178         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 5 columns]

In [16]:
df.drop(['level_0','index',], axis = 'columns', inplace = True)    #cleaning the data more.
df


Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [17]:
df.shape

(103, 3)

**We have 103 rows and 3 columns after getting the final pandas dataframe.**

##### **_This concludes part 1._**

# Part 2:

In [18]:
#getting the latitude and longitude data from csv file
ll_data = pd.read_csv('https://cocl.us/Geospatial_data')
ll_data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

Since our data frame df and ll_data both have same number of rows and columns, we need to merge them based on postal address.

In [19]:
combined_data = df.join(ll_data.set_index('Postal Code'), on='Postal Code', how='inner')
combined_data

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [20]:
combined_data.shape

(103, 5)

**We have achieved the required dataset.**

#### **_This concludes part 2._**

# Part 3:

Using geopy library to get the latitude and longitude values of Toronto city.


In [24]:
from geopy.geocoders import Nominatim


In [22]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


Creating the map using latitude and longitude values

In [25]:
!pip install folium
import folium

     |████████████████████████████████| 94 kB 3.4 MB/s eta 0:00:011


In [26]:
# Creating the map of Toronto
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for latitude, longitude, borough, neighbourhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

Initializing Foursquare API credentials

In [27]:
CLIENT_ID = '1500GB54ZDMYDQEQ2YYD10WFUQZQQ4IGFL2IBU1WBEPPE0WT' 
CLIENT_SECRET = 'AIXXCYMHUA4WXHNUT2B5YBEEKJ5HYDRVGILAT03RIVDYUNRB'
VERSION = '20180605' 
LIMIT = 100 

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 1500GB54ZDMYDQEQ2YYD10WFUQZQQ4IGFL2IBU1WBEPPE0WT
CLIENT_SECRET:AIXXCYMHUA4WXHNUT2B5YBEEKJ5HYDRVGILAT03RIVDYUNRB


### Now we will create a function to explore neighborhoods in toronto:

In [28]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

Collecting the venues in Toronto for each Neighbourhood:

In [30]:
venues_in_toronto = getNearbyVenues(combined_data['Neighbourhood'], combined_data['Latitude'], combined_data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

Checking the data:

In [31]:
venues_in_toronto.shape

(1329, 5)

In [32]:
venues_in_toronto.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2         Parkwoods               43.753259               -79.329656   
3  Victoria Village               43.725882               -79.315572   
4  Victoria Village               43.725882               -79.315572   

                    Venue         Venue Category  
0         Brookbanks Park                   Park  
1     Towns On The Ravine                  Hotel  
2           Variety Store      Food & Drink Shop  
3  Victoria Village Arena           Hockey Arena  
4               Portugril  Portuguese Restaurant

Checking the Venues based on Neighbourhood:

In [33]:
venues_in_toronto.groupby('Neighbourhood').head()

Neighbourhood  \
0                                             Parkwoods   
1                                             Parkwoods   
2                                             Parkwoods   
3                                      Victoria Village   
4                                      Victoria Village   
...                                                 ...   
1313  Mimico NW, The Queensway West, South of Bloor,...   
1314  Mimico NW, The Queensway West, South of Bloor,...   
1315  Mimico NW, The Queensway West, South of Bloor,...   
1316  Mimico NW, The Queensway West, South of Bloor,...   
1317  Mimico NW, The Queensway West, South of Bloor,...   

      Neighbourhood Latitude  Neighbourhood Longitude                   Venue  \
0                  43.753259               -79.329656         Brookbanks Park   
1                  43.753259               -79.329656     Towns On The Ravine   
2                  43.753259               -79.329656           Variety Store   
3                  43.725882               -79.315572  Victoria Village Arena   
4                  43.725882               -79.315572               Portugril   
...                      ...                      ...                     ...   
1313               43.628841               -79.520999              Wingporium   
1314               43.628841               -79.520999        South St. Burger   
1315               43.628841               -79.520999               Dollarama   
1316               43.628841               -79.520999          Healthy Planet   
1317               43.628841               -79.520999    Artisano Bakery Café   

             Venue Category  
0                      Park  
1                     Hotel  
2         Food & Drink Shop  
3              Hockey Arena  
4     Portuguese Restaurant  
...                     ...  
1313            Wings Joint  
1314           Burger Joint  
1315         Discount Store  
1316        Supplement Shop  
1317                 Bakery  

[418 rows x 5 columns]

So there are 405 records for each neighbourhood.

Checking for the maximum venue categories:

In [34]:
venues_in_toronto.groupby('Venue Category').max()

Neighbourhood  \
Venue Category                                                          
Accessories Store                                   Wexford, Maryvale   
Adult Boutique                                   Church and Wellesley   
Airport                                                     Downsview   
Airport Food Court  CN Tower, King and Spadina, Railway Lands, Har...   
Airport Gate        CN Tower, King and Spadina, Railway Lands, Har...   
...                                                               ...   
Wine Bar                     Toronto Dominion Centre, Design Exchange   
Wine Shop                                Dufferin, Dovercourt Village   
Wings Joint         Mimico NW, The Queensway West, South of Bloor,...   
Women's Store                                     Caledonia-Fairbanks   
Yoga Studio                            University of Toronto, Harbord   

                    Neighbourhood Latitude  Neighbourhood Longitude  \
Venue Category                                                        
Accessories Store                43.750072               -79.295849   
Adult Boutique                   43.665860               -79.383160   
Airport                          43.737473               -79.394420   
Airport Food Court               43.628947               -79.394420   
Airport Gate                     43.628947               -79.394420   
...                                    ...                      ...   
Wine Bar                         43.653206               -79.379817   
Wine Shop                        43.669005               -79.442259   
Wings Joint                      43.628841               -79.520999   
Women's Store                    43.689026               -79.453512   
Yoga Studio                      43.715383               -79.340923   

                                              Venue  
Venue Category                                       
Accessories Store                       Puffin Gear  
Adult Boutique                            Seduction  
Airport             Toronto Downsview Airport (YZD)  
Airport Food Court                Billy Bishop Café  
Airport Gate                                 Gate 8  
...                                             ...  
Wine Bar                          The National Club  
Wine Shop                                    Macedo  
Wings Joint                              Wingporium  
Women's Store                         Maximum Woman  
Yoga Studio                               YogaSpace  

[236 rows x 4 columns]

There are around 232 different types of Venue Categories.

We can do a lot of stuff as was mentioned in the lab like calculating the mean venue categories in each Neighbourhood,one hot encoding and so on.


#### One Hot encoding the venue Categories:

In [40]:
toronto_venue_cat = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_venue_cat

Accessories Store  Adult Boutique  Airport  Airport Food Court  \
0                     0               0        0                   0   
1                     0               0        0                   0   
2                     0               0        0                   0   
3                     0               0        0                   0   
4                     0               0        0                   0   
...                 ...             ...      ...                 ...   
1324                  0               0        0                   0   
1325                  0               0        0                   0   
1326                  0               0        0                   0   
1327                  0               0        0                   0   
1328                  0               0        0                   0   

      Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0                0               0                0                 0   
1                0               0                0                 0   
2                0               0                0                 0   
3                0               0                0                 0   
4                0               0                0                 0   
...            ...             ...              ...               ...   
1324             0               0                0                 0   
1325             0               0                0                 0   
1326             0               0                0                 0   
1327             0               0                0                 0   
1328             0               0                0                 0   

      American Restaurant  Antique Shop  ...  Vegetarian / Vegan Restaurant  \
0                       0             0  ...                              0   
1                       0             0  ...                              0   
2                       0             0  ...                              0   
3                       0             0  ...                              0   
4                       0             0  ...                              0   
...                   ...           ...  ...                            ...   
1324                    0             0  ...                              0   
1325                    0             0  ...                              0   
1326                    0             0  ...                              0   
1327                    0             0  ...                              0   
1328                    0             0  ...                              0   

      Video Game Store  Video Store  Vietnamese Restaurant  Warehouse Store  \
0                    0            0                      0                0   
1                    0            0                      0                0   
2                    0            0                      0                0   
3                    0            0                      0                0   
4                    0            0                      0                0   
...                ...          ...                    ...              ...   
1324                 0            0                      0                0   
1325                 0            0                      0                0   
1326                 0            0                      0                0   
1327                 0            0                      0                0   
1328                 0            0                      0                0   

      Wine Bar  Wine Shop  Wings Joint  Women's Store  Yoga Studio  
0            0          0            0              0            0  
1            0          0            0              0            0  
2            0          0            0              0            0  
3            0          0            0              0            0  
4            0       

Adding the neighbourhood to the encoded dataframe:

In [41]:
toronto_venue_cat['Neighbourhood'] = venues_in_toronto['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [toronto_venue_cat.columns[-1]] + list(toronto_venue_cat.columns[:-1])
toronto_venue_cat = toronto_venue_cat[fixed_columns]

toronto_venue_cat.head()

Neighbourhood  Accessories Store  Adult Boutique  Airport  \
0         Parkwoods                  0               0        0   
1         Parkwoods                  0               0        0   
2         Parkwoods                  0               0        0   
3  Victoria Village                  0               0        0   
4  Victoria Village                  0               0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  ...  Vegetarian / Vegan Restaurant  \
0                 0                    0  ...                              0   
1                 0                    0  ...                              0   
2                 0                    0  ...                              0   
3                 0                    0  ...                              0   
4                 0                    0  ...                              0   

   Video Game Store  Video Store  Vietnamese Restaurant  Warehouse Store  \
0                 0            0                      0                0   
1                 0            0                      0                0   
2                 0            0                      0                0   
3                 0            0                      0                0   
4                 0            0                      0                0   

   Wine Bar  Wine Shop  Wings Joint  Women's Store  Yoga Studio  
0         0          0            0              0            0  
1         0          0            0              0            0  
2         0          0            0              0            0  
3         0          0            0              0            0  
4         0          0            0              0            0  

[5 rows x 237 columns]

We will group the Neighbourhoods, calculate the mean venue categories in each Neighbourhood:

In [42]:
toronto_grouped = toronto_venue_cat.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Adult Boutique  Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0             0.0      0.0                 0.0           0.0             0.0   
1             0.0      0.0                 0.0           0.0             0.0   
2             0.0      0.0                 0.0           0.0             0.0   
3             0.0      0.0                 0.0           0.0             0.0   
4             0.0      0.0                 0.0           0.0             0.0   

   Airport Service  Airport Terminal  American Restaurant  ...  \
0              0.0               0.0             0.000000  ...   
1              0.0               0.0             0.000000  ...   
2              0.0               0.0             0.000000  ...   
3              0.0               0.0             0.000000  ...   
4              0.0               0.0             0.043478  ...   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                            0.0               0.0          0.0   
1                            0.0               0.0          0.0   
2                            0.0               0.0          0.0   
3                            0.0               0.0          0.0   
4                            0.0               0.0          0.0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wine Shop  Wings Joint  \
0                    0.0              0.0       0.0        0.0          0.0   
1                    0.0              0.0       0.0        0.0          0.0   
2                    0.0              0.0       0.0        0.0          0.0   
3                    0.0              0.0       0.0        0.0          0.0   
4                    0.0              0.0       0.0        0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  

[5 rows x 237 columns]

Now,

Let's make a function to get the top most common venue categories:

In [35]:

def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

## Let's move to clustering the neighborhoods now:

There are way too many venue categories, we can take the top 10 to cluster the neighbourhoods:

In [37]:
import numpy as np

In [43]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village                  Café   
4                Bedford Park, Lawrence Manor East    Italian Restaurant   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0  Latin American Restaurant        Breakfast Spot        Clothing Store   
1                        Gym          Skating Rink           Coffee Shop   
2                       Bank              Pharmacy         Deli / Bodega   
3                       Bank    Chinese Restaurant   Japanese Restaurant   
4                  Juice Bar           Coffee Shop        Sandwich Place   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0          Skating Rink           Yoga Studio    Dim Sum Restaurant   
1                  Pool                   Pub              Pharmacy   
2           Bridal Shop         Shopping Mall                 Diner   
3    Dim Sum Restaurant    Falafel Restaurant           Event Space   
4              Pharmacy      Sushi Restaurant     Indian Restaurant   

   8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
0   Ethiopian Restaurant     Electronics Store  Eastern European Restaurant  
1  General Entertainment               Gay Bar                      Dog Run  
2         Sandwich Place           Supermarket             Sushi Restaurant  
3   Ethiopian Restaurant     Electronics Store  Eastern European Restaurant  
4                   Café               Butcher              Thai Restaurant

Let's make the model to cluster our Neighbourhoods:

In [44]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [45]:
# set number of clusters
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans


KMeans(n_clusters=5, random_state=0)

In [47]:
kmeans.labels_[0:100]   #checking labels

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1,
       1, 1, 0, 1, 1, 3, 0, 1, 0, 1, 1, 1, 1, 1, 2, 1, 1, 0, 1, 1, 1, 0,
       1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 4, 1,
       0, 1, 1, 1, 1, 1, 0], dtype=int32)

Adding the clustering Label column to the top 10 common venue categories:

In [ ]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Join toronto_grouped with combined_data on neighbourhood to add latitude & longitude for each neighborhood to prepare it for plotting:

In [53]:
toronto_merged = combined_data

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656             0.0                  Park   
1  43.725882 -79.315572             1.0           Pizza Place   
2  43.654260 -79.360636             1.0           Coffee Shop   
3  43.718518 -79.464763             1.0        Clothing Store   
4  43.662301 -79.389494             1.0           Coffee Shop   

  2nd Most Common Venue  3rd Most Common Venue 4th Most Common Venue  \
0     Food & Drink Shop                  Hotel           Yoga Studio   
1           Coffee Shop  Portuguese Restaurant          Hockey Arena   
2                  Park                 Bakery               Theater   
3     Accessories Store  Vietnamese Restaurant              Boutique   
4      Sushi Restaurant            Yoga Studio   Fried Chicken Joint   

  5th Most Common Venue   6th Most Common Venue 7th Most Common Venue  \
0          Dessert Shop             Event Space  Ethiopian Restaurant   
1  Ethiopian Restaurant             Event Space     Electronics Store   
2        Breakfast Spot   Performing Arts Venue        Chocolate Shop   
3    Miscellaneous Shop  Furniture / Home Store           Event Space   
4                  Bank      Italian Restaurant                  Café   

         8th Most Common Venue        9th Most Common Venue  \
0            Electronics Store  Eastern European Restaurant   
1  Eastern European Restaurant             Department Store   
2                          Pub                   Restaurant   
3                  Coffee Shop               Cosmetics Shop   
4           Mexican Restaurant                        Diner   

  10th Most Common Venue  
0              Drugstore  
1              Drugstore  
2                   Café  
3   Ethiopian Restaurant  
4         Discount Store

In [54]:
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

## Plotting the clusters on the map:

In [58]:
from matplotlib import *

In [59]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighbourhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

### We can verify our clusters:


In [60]:
#cluster 1
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0         North York             0.0                  Park   
21              York             0.0                  Park   
32       Scarborough             0.0            Playground   
35         East York             0.0                  Park   
61   Central Toronto             0.0                  Park   
64              York             0.0                  Park   
66        North York             0.0                  Park   
83   Central Toronto             0.0            Playground   
85       Scarborough             0.0            Playground   
91  Downtown Toronto             0.0                  Park   
98         Etobicoke             0.0                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0      Food & Drink Shop                 Hotel           Yoga Studio   
21         Women's Store                  Pool      Department Store   
32      Department Store           Event Space  Ethiopian Restaurant   
35     Convenience Store           Yoga Studio          Dessert Shop   
61           Swim School              Bus Line           Yoga Studio   
64     Convenience Store           Yoga Studio          Dessert Shop   
66     Convenience Store                   Bar           Yoga Studio   
83                 Trail      Department Store           Event Space   
85          Intersection                  Park           Event Space   
91            Playground                 Trail         Deli / Bodega   
98                 River           Yoga Studio      Department Store   

   5th Most Common Venue        6th Most Common Venue  \
0           Dessert Shop                  Event Space   
21           Event Space         Ethiopian Restaurant   
32     Electronics Store  Eastern European Restaurant   
35           Event Space         Ethiopian Restaurant   
61    Dim Sum Restaurant                  Event Space   
64           Event Space         Ethiopian Restaurant   
66    Dim Sum Restaurant           Falafel Restaurant   
83  Ethiopian Restaurant            Electronics Store   
85  Ethiopian Restaurant            Electronics Store   
91  Ethiopian Restaurant            Electronics Store   
98           Event Space         Ethiopian Restaurant   

          7th Most Common Venue        8th Most Common Venue  \
0          Ethiopian Restaurant            Electronics Store   
21            Electronics Store  Eastern European Restaurant   
32                    Drugstore                   Donut Shop   
35            Electronics Store  Eastern European Restaurant   
61         Ethiopian Restaurant            Electronics Store   
64            Electronics Store  Eastern European Restaurant   
66                  Event Space         Ethiopian Restaurant   
83  Eastern European Restaurant                    Drugstore   
85  Eastern European Restaurant                    Drugstore   
91  Eastern European Restaurant                    Drugstore   
98            Electronics Store  Eastern European Restaurant   

          9th Most Common Venue       10th Most Common Venue  
0   Eastern European Restaurant                    Drugstore  
21                    Drugstore                   Donut Shop  
32                      Dog Run          Distribution Center  
35                    Drugstore                   Donut Shop  
61  Eastern European Restaurant                    Drugstore  
64                    Drugstore                   Donut Shop  
66            Electronics Store  Eastern European Restaurant  
83                   Donut Shop                      Dog Run  
85                   Donut Shop                      Dog Run  
91                   Donut Shop                      Dog Run  
98                    Drugstore                   Donut Shop

In [61]:
#cluster 2
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
1          North York             1.0           Pizza Place   
2    Downtown Toronto             1.0           Coffee Shop   
3          North York             1.0        Clothing Store   
4    Downtown Toronto             1.0           Coffee Shop   
7          North York             1.0                   Gym   
..                ...             ...                   ...   
96   Downtown Toronto             1.0            Restaurant   
97   Downtown Toronto             1.0                  Café   
99   Downtown Toronto             1.0           Coffee Shop   
100      East Toronto             1.0    Light Rail Station   
102         Etobicoke             1.0  Fast Food Restaurant   

    2nd Most Common Venue  3rd Most Common Venue 4th Most Common Venue  \
1             Coffee Shop  Portuguese Restaurant          Hockey Arena   
2                    Park                 Bakery               Theater   
3       Accessories Store  Vietnamese Restaurant              Boutique   
4        Sushi Restaurant            Yoga Studio   Fried Chicken Joint   
7             Coffee Shop             Restaurant            Beer Store   
..                    ...                    ...                   ...   
96                 Bakery                   Café    Italian Restaurant   
97             Restaurant            Coffee Shop    Seafood Restaurant   
99       Sushi Restaurant        Thai Restaurant      Ramen Restaurant   
100                  Park             Restaurant  Gym / Fitness Center   
102        Discount Store                 Bakery       Supplement Shop   

    5th Most Common Venue   6th Most Common Venue   7th Most Common Venue  \
1    Ethiopian Restaurant             Event Space       Electronics Store   
2          Breakfast Spot   Performing Arts Venue          Chocolate Shop   
3      Miscellaneous Shop  Furniture / Home Store             Event Space   
4                    Bank      Italian Restaurant                    Café   
7     Japanese Restaurant      Italian Restaurant          Clothing Store   
..                    ...                     ...                     ...   
96            Coffee Shop     Japanese Restaurant    Caribbean Restaurant   
97                  Hotel          Sandwich Place                     Pub   
99         Ice Cream Shop       Indian Restaurant                   Diner   
100           Pizza Place              Comic Shop                 Brewery   
102         Burrito Place            Burger Joint  Thrift / Vintage Store   

           8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
1    Eastern European Restaurant      Department Store              Drugstore  
2                            Pub            Restaurant                   Café  
3                    Coffee Shop        Cosmetics Shop   Ethiopian Restaurant  
4             Mexican Restaurant                 Diner         Discount Store  
7             Chinese Restaurant  Caribbean Restaurant     Dim Sum Restaurant  
..                           ...                   ...                    ...  
96          Taiwanese Restaurant                  Bank              Gift Shop  
97                        Bakery             Speakeasy            Pizza Place  
99                    Restaurant   Japanese Restaurant                   Café  
100                Auto Workshop        Farmers Market   Fast Food Restaurant  
102               Sandwich Place            Kids Store          Tanning Salon  

[84 rows x 12 columns]

In [62]:
#cluster 3
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
57   North York             2.0        Baseball Field           Yoga Studio   
101   Etobicoke             2.0        Baseball Field           Yoga Studio   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
57   Fast Food Restaurant                  Farm    Falafel Restaurant   
101  Fast Food Restaurant                  Farm    Falafel Restaurant   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
57            Event Space  Ethiopian Restaurant     Electronics Store   
101           Event Space  Ethiopian Restaurant     Electronics Store   

           9th Most Common Venue 10th Most Common Venue  
57   Eastern European Restaurant              Drugstore  
101  Eastern European Restaurant              Drugstore

In [63]:
#cluster 4
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
6  Scarborough             3.0  Fast Food Restaurant          Dessert Shop   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
6    Falafel Restaurant           Event Space  Ethiopian Restaurant   

  6th Most Common Venue        7th Most Common Venue 8th Most Common Venue  \
6     Electronics Store  Eastern European Restaurant             Drugstore   

  9th Most Common Venue 10th Most Common Venue  
6            Donut Shop                Dog Run

In [64]:
#cluster 5
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
11  Etobicoke             4.0                Bakery           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
11  Fast Food Restaurant                  Farm    Falafel Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
11           Event Space  Ethiopian Restaurant     Electronics Store   

          9th Most Common Venue 10th Most Common Venue  
11  Eastern European Restaurant              Drugstore

we have successfully clustered toronto based on venue categories!

## _This conludes part 3 and this assignment._

